scheduler/core/components/optimizer/timeline.py

```diff
@@ -25,11 +26,109 @@ class Timeline:
     _total_time_slots: int
 
     def __post_init__(self):
-        self.schedule = np.full(self._total_time_slots, EMPTY)
+        self.time_slots = np.full(self._total_time_slots, EMPTY)
         self.is_full = False
 
     def __contains__(self, obs: Observation) -> bool:
-        return any(visit.obs_id == obs.id for visit in self.visits)
+        return any(obs.id in self.time_slots)
+
+    def _empty_slots(self) -> np.ndarray:
+        """
+        Determine the number of empty time slots..
+        """
+        return np.where(self.time_slots == EMPTY)[0][:]
+
+    def _available_intervals(self, empty_slots: np.ndarray) -> np.ndarray:
+        """
+        Calculate the available intervals in the schedule by creating an array that contains all
+        the groups of consecutive numbers
+        """
+        return np.split(empty_slots, np.where(np.diff(empty_slots) != 1)[0]+1)
+
+    def get_available_intervals(self, first) -> np.ndarray:
+        """
+        Get the set of time_slot intervals that can be scheduled. If desired, return only the first
+        """
+        empty_slots = self._empty_slots()
+        intervals = self._available_intervals(empty_slots)
+        # print(f"length intervals = {len(intervals)}")
+        return intervals[0] if first and len(intervals) > 1 else intervals
+
+    def get_earliest_available_interval(self) -> np.ndarray:
+        """
+        Get the earliest available space in the schedule that can allocate an observation
+        """
+        return self.get_available_intervals()[0]
+
+    def add(self, iobs: int, time_slots: int, interval: int) -> datetime:
+        """
+        Add an observation index to the first open position (-1) in the given interval
+        Returns the time of this position
+        """
+        # interval = self.get_earliest_available_interval()
+        # Get first non-zero slot in given interval
+        i_first = np.where(self.time_slots[interval] == EMPTY)[0][0]
+
+        # Set values of time_slots to the observation index
+        self.time_slots[interval[i_first:i_first + time_slots]] = iobs
+
+        # Clock time for the starting index
+        start = self.start + interval[i_first] * self.time_slot_length
```

Comment:
This will crash, because `interval` is specified as an `int` and you are trying to index it.

I think you might mean that `interval` is an `npt.NDArray[int]`?

In the interests of saving time and typing, I would recommend typealiasing at the top after the imports:

```python
Interval = npt.NDarray[int]
```

Reply:
See my previous comment... I'm moving the type alias to this to the `optimizer` `__init__.py` file since we should be using it in the `BaseOptimizer.add` method like you do here. 

Once that's done, you can then import it with:
```python
from . import Interval
```